### ISSUE TITLE
Can't upload empty file when using asyncReadFileFn hook [v3 regression]

### ISSUE THREAD

**Issue by bertrandg** [flowjs/flow.js#375]
Hi,

I'm using V3 branch since multiple months and it works well but i've discoved this small regression.
I'm unable to send empty files when using `asyncReadFileFn` hook.

Investigating, it seems to come from there:
https://github.com/flowjs/flow.js/blob/5591263e71426fb811f9495457380f81798a09af/src/FlowChunk.js#L341

From what i've seen convert to this fix it:
`if (data && (data.size > 0 || this.fileObj.size === 0)) {`

But i would like to have a feedback from you @drzraf before making an MR.

**Comment by AidasK** [comment]
Maybe you could create a failing test to demonstrate this?

**Comment by drzraf** [comment]
From a quick look, the problem is that we must determinate between an empty **regular** file and a stream which finished reading (thus returning 0)

Why are you using the `asyncRead` feature/codepath with a simple regular file (for which I believe most processing hook could be synchronous)?

**Comment by bertrandg** [comment]
@AidasK yes I will try to add one.

@drzraf 
I'm not sure to understand, i'm using the asyncReadFileFn hook when I create my flowjs instance to being able to generate a sha1 hash of the chunk 'on the fy' and it works well except for empty file sadly, I'm not using stream..

**Comment by drzraf** [comment]
Quoting from https://github.com/flowjs/flow.js/issues/353#issuecomment-1028498996

> #354 (#363) broke the testsuite but more importantly cause possible problem regarding `asyncReadFileFn` (see #368)

So the code was put in a instable stable at that commit.

This was a huge commit, it changed the semantics, the regression wasn't spotted soon enough but, more importantly, fixing it implies deeply architectural changes. (The problem was that going full-async would break the behavior inside xhr.request listeners and the recursion mechanism)

I'm sorry I didn't have the time to dig into this again more than when I stopped at #368 (when I understood that relying upon XHR recursive handlers was inherently incompatible with the _async_ changes already merged in #363)

**Comment by drzraf** [comment]
You're right. dist/* were not regenerated (and committed) lately. Will consider that if we ever fix #368
